lib/puppet/parser/compiler.rb

```diff
@@ -865,11 +865,32 @@ def create_settings_scope
     scope = @topscope.class_scope(settings_type)
 
     env = environment
+    settings_hash = {}
     Puppet.settings.each do |name, setting|
       next if name == :name
-      scope[name.to_s] = env[name]
+      name = name.to_s
+      # Construct a hash (in anticipation it will be set in top scope under a name like $settings)
+      settings_hash[name] = transform_setting(env[name])
```

Comment:
Converting `name` to a `String`  before passing it to `env[name]` is inefficient. The first thing that happens in `settings#value()` is that it is converted back to a `Symbol`. I suggest keeping the behavior of passing it as a symbol.